File: niko-studio/src/knowledge/services/manager.py
"""
服务管理器

统一管理 LLM 和 Embedding 服务的生命周期，支持健康检查和热重载。
"""

import asyncio
from typing import Any

from .models import (
    ProviderConfig,
    ProviderType,
    ServiceConfig,
)
from .cache import InMemoryEmbeddingCache
from .llm_service import LLMServiceImpl
from .embedding_service import EmbeddingServiceImpl
from .providers import (
    OpenAILLMProvider,
    AnthropicLLMProvider,
    OpenAIEmbeddingProvider,
    LocalEmbeddingProvider,
)


class ServiceManager:
    """服务管理器

    单例模式管理所有 LLM/Embedding 服务实例，支持:
    - 服务初始化和关闭
    - 健康检查
    - 配置热重载
    """

    _instance: "ServiceManager | None" = None

    def __new__(cls, *args, **kwargs):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def __init__(self, config: ServiceConfig | None = None):
        """初始化服务管理器

        Args:
            config: 服务配置
        """
        if hasattr(self, "_initialized") and self._initialized:
            return

        self._config = config or ServiceConfig()
        self._llm_providers: dict[ProviderType, Any] = {}
        self._embedding_providers: dict[ProviderType, Any] = {}
        self._llm_service: LLMServiceImpl | None = None
        self._embedding_service: EmbeddingServiceImpl | None = None
        self._cache: InMemoryEmbeddingCache | None = None
        self._health_status: dict[str, bool] = {}
        self._health_check_task: asyncio.Task | None = None
        self._initialized = False

    async def initialize(self) -> None:
        """初始化所有服务"""
        if self._initialized:
            return

        # 初始化缓存
        if self._config.embedding_cache_enabled:
            self._cache = InMemoryEmbeddingCache(
                max_size=self._config.embedding_cache_max_size,
                default_ttl=self._config.embedding_cache_ttl,
            )

        # 初始化 Providers
        for provider_config in self._config.providers:
            await self._init_provider(provider_config)

        # 初始化服务
        self._llm_service = LLMServiceImpl(
            providers=self._llm_providers,
            default_provider=self._config.default_llm_provider,
            max_retries=self._config.retry_max_attempts,
            retry_base_delay=self._config.retry_initial_delay,
        )

        self._embedding_service = EmbeddingServiceImpl(
            providers=self._embedding_providers,
            default_provider=self._config.default_embedding_provider,
            cache=self._cache,
        )

        # 启动健康检查
        self._health_check_task = asyncio.create_task(self._health_check_loop())
        self._initialized = True

    async def _init_provider(self, config: ProviderConfig) -> None:
        """初始化单个 Provider"""
        ptype = config.provider

        if ptype == ProviderType.OPENAI:
            self._llm_providers[ptype] = OpenAILLMProvider(
                api_key=config.api_key,
                base_url=config.base_url,
                organization=config.organization,
                model_mapping=config.model_mapping,
                timeout=config.timeout,
                max_retries=config.max_retries,
            )
            self._embedding_providers[ptype] = OpenAIEmbeddingProvider(
                api_key=config.api_key,
                base_url=config.base_url,
                organization=config.organization,
                default_model=config.embedding_model,
                timeout=config.timeout,
                max_retries=config.max_retries,
            )

        elif ptype == ProviderType.ANTHROPIC:
            self._llm_providers[ptype] = AnthropicLLMProvider(
                api_key=config.api_key,
                base_url=config.base_url,
                model_mapping=config.model_mapping,
                timeout=config.timeout,
                max_retries=config.max_retries,
            )

        elif ptype == ProviderType.LOCAL:
            self._embedding_providers[ptype] = LocalEmbeddingProvider(
                model_name=config.embedding_model,
                backend=config.base_url or "fastembed",
            )

    async def shutdown(self) -> None:
        """关闭所有服务"""
        if self._health_check_task:
            self._health_check_task.cancel()
            try:
                await self._health_check_task
            except asyncio.CancelledError:
                pass

        if self._cache:
            await self._cache.clear()

        self._llm_providers.clear()
        self._embedding_providers.clear()
        self._initialized = False
        ServiceManager._instance = None

    @property
    def llm(self) -> LLMServiceImpl:
        """获取 LLM 服务"""
        if not self._llm_service:
            raise RuntimeError("ServiceManager not initialized. Call initialize() first.")
        return self._llm_service

    @property
    def embedding(self) -> EmbeddingServiceImpl:
        """获取 Embedding 服务"""
        if not self._embedding_service:
            raise RuntimeError("ServiceManager not initialized. Call initialize() first.")
        return self._embedding_service

    async def check_health(self) -> dict[str, bool]:
        """检查所有 Provider 健康状态"""
        checks: dict[str, bool] = {}

        for ptype, provider in self._llm_providers.items():
            checks[f"llm_{ptype.value}"] = await provider.health_check()

        for ptype, provider in self._embedding_providers.items():
            checks[f"embedding_{ptype.value}"] = await provider.health_check()

        self._health_status = checks
        return checks

    async def _health_check_loop(self) -> None:
        """健康检查循环"""
        while True:
            try:
                await asyncio.sleep(self._config.health_check_interval)
                await self.check_health()
            except asyncio.CancelledError:
                break
            except Exception:
                pass

    def is_healthy(self) -> bool:
        """检查服务是否健康"""
        return any(self._health_status.values())

    def get_health_status(self) -> dict[str, bool]:
        """获取健康状态"""
        return self._health_status.copy()

    async def get_cache_stats(self) -> dict[str, Any] | None:
        """获取缓存统计"""
        if self._cache:
            return await self._cache.stats()
        return None

File: src-ts/knowledge/manager.ts
/**
 * Knowledge module - service manager
 *
 * Manages the lifecycle of LLM and Embedding services.
 * Singleton pattern with health checking and hot-reload support.
 */

import type {
  ProviderConfig,
  ServiceConfig,
  LLMProvider as LLMServiceProviderContract,
  EmbeddingProvider as EmbeddingServiceProviderContract,
  LLMResponse,
  EmbeddingResponse,
  StreamChunk,
} from './models';
import {
  ProviderType,
  createServiceConfig,
} from './models';
import { TieredEmbeddingCache } from './cache';
import { LLMServiceImpl } from './llm-service';
import { EmbeddingServiceImpl } from './embedding-service';
import {
  OpenAILLMProvider,
  AnthropicLLMProvider,
  OpenAIEmbeddingProvider,
  LocalEmbeddingProvider,
} from './providers';

const DEFAULT_LOCAL_EMBEDDING_MODEL = 'BAAI/bge-small-en-v1.5';

interface LLMRuntimeProvider {
  readonly providerType: string;
  getModelForTier(tier: string): string;
  complete(
    prompt: string,
    model: string,
    options?: {
      temperature?: number;
      maxTokens?: number;
      systemPrompt?: string;
      stopSequences?: string[];
      responseFormat?: Record<string, unknown>;
    },
  ): Promise<LLMResponse>;
  streamComplete(
    prompt: string,
    model: string,
    options?: {
      temperature?: number;
      maxTokens?: number;
      systemPrompt?: string;
      stopSequences?: string[];
    },
  ): AsyncGenerator<StreamChunk, void, undefined>;
  healthCheck(): Promise<boolean>;
}

interface EmbeddingRuntimeProvider {
  readonly providerType: string;
  getDimensions(model: string): number;
  embed(
    texts: string[],
    model: string,
    options?: { dimensions?: number },
  ): Promise<EmbeddingResponse>;
  healthCheck(): Promise<boolean>;
}

function createLLMServiceProvider(
  provider: LLMRuntimeProvider,
): LLMServiceProviderContract {
  return {
    providerType: provider.providerType,
    getModelForTier: (tier) => provider.getModelForTier(tier),
    generate: async (request) => {
      const model = request.modelOverride ?? provider.getModelForTier(request.modelTier);
      return provider.complete(request.prompt, model, {
        temperature: request.temperature,
        maxTokens: request.maxTokens ?? undefined,
        systemPrompt: request.systemPrompt ?? undefined,
        stopSequences: request.stopSequences,
        responseFormat: request.responseFormat ?? undefined,
      });
    },
    streamGenerate: async function* (request) {
      const model = request.modelOverride ?? provider.getModelForTier(request.modelTier);
      for await (const chunk of provider.streamComplete(request.prompt, model, {
        temperature: request.temperature,
        maxTokens: request.maxTokens ?? undefined,
        systemPrompt: request.systemPrompt ?? undefined,
        stopSequences: request.stopSequences,
      })) {
        yield chunk;
      }
    },
    healthCheck: () => provider.healthCheck(),
  };
}

function createEmbeddingServiceProvider(
  provider: EmbeddingRuntimeProvider,
): EmbeddingServiceProviderContract {
  return {
    providerType: provider.providerType,
    getModel: () => '',
    embed: async (texts, model) => {
      const response = await provider.embed(texts, model ?? '');
      return response.embeddings;
    },
    healthCheck: () => provider.healthCheck(),
  };
}

/**
 * Service manager
 *
 * Singleton that manages all LLM/Embedding service instances.
 * Supports:
 * - Service initialization and shutdown
 * - Health checking
 * - Configuration hot-reload
 */
export class ServiceManager {
  private static _instance: ServiceManager | null = null;

  private _config: ServiceConfig;
  private readonly _llmProviders: Map<string, LLMServiceProviderContract>;
  private readonly _embeddingProviders: Map<string, EmbeddingServiceProviderContract>;
  private _llmService: LLMServiceImpl | null = null;
  private _embeddingService: EmbeddingServiceImpl | null = null;
  private _cache: TieredEmbeddingCache | null = null;
  private _healthStatus: Record<string, boolean> = {};
  // 按需健康检查：不再用 setInterval 轮询，仅在请求失败时触发
  private _lastHealthCheckMs: number = 0;
  private readonly _minHealthCheckIntervalMs: number = 60_000; // 最小 60s 间隔防抖
  private _initialized: boolean = false;

  private constructor(config?: ServiceConfig | null) {
    this._config = config ?? createServiceConfig();
    this._llmProviders = new Map();
    this._embeddingProviders = new Map();
  }

  /**
   * Get the singleton instance
   */
  static getInstance(config?: ServiceConfig | null): ServiceManager {
    if (!ServiceManager._instance) {
      ServiceManager._instance = new ServiceManager(config);
    }
    return ServiceManager._instance;
  }

  /**
   * Reset the singleton (for testing)
   */
  static resetInstance(): void {
    if (ServiceManager._instance) {
      ServiceManager._instance.shutdown();
    }
    ServiceManager._instance = null;
  }

  /**
   * Initialize all services
   */
  async initialize(): Promise<void> {
    if (this._initialized) return;

    // Initialize cache
    if (this._config.embeddingCacheEnabled) {
      this._cache = new TieredEmbeddingCache(
        this._config.embeddingCacheMaxSize,
        this._config.embeddingCacheTTL,
      );
    }

    // Initialize providers
    for (const providerConfig of this._config.providers) {
      await this._initProvider(providerConfig);
    }

    if (
      this._config.defaultEmbeddingProvider === ProviderType.LOCAL
      && !this._embeddingProviders.has(ProviderType.LOCAL)
    ) {
      const embeddingProvider = new LocalEmbeddingProvider({
        modelName: DEFAULT_LOCAL_EMBEDDING_MODEL,
        backend: 'fastembed',
      });
      this._embeddingProviders.set(ProviderType.LOCAL, createEmbeddingServiceProvider(embeddingProvider));
    }

    // Initialize services
    this._llmService = new LLMServiceImpl({
      providers: this._llmProviders,
      defaultProvider: this._config.defaultLLMProvider,
      maxRetries: this._config.retryMaxAttempts,
      retryBaseDelay: this._config.retryInitialDelay,
    });

    this._embeddingService = new EmbeddingServiceImpl({
      providers: this._embeddingProviders,
      defaultProvider: this._config.defaultEmbeddingProvider,
      cache: this._cache,
    });

    // 初始化时执行一次健康检查，之后仅在请求失败时按需触发
    this.checkHealth().catch(() => {
      // 初始化阶段的健康检查失败不影响启动
    });

    this._initialized = true;
  }

  /**
   * Initialize a single provider
   */
  private async _initProvider(config: ProviderConfig): Promise<void> {
    const ptype = config.provider;

    if (ptype === ProviderType.OPENAI) {
      const llmProvider = new OpenAILLMProvider({
        apiKey: config.apiKey,
        baseUrl: config.baseUrl,
        organization: config.organization,
        modelMapping: config.modelMapping,
        timeout: config.timeout,
        maxRetries: config.maxRetries,
      });
      this._llmProviders.set(ptype, createLLMServiceProvider(llmProvider));

      const embeddingProvider = new OpenAIEmbeddingProvider({
        apiKey: config.apiKey,
        baseUrl: config.baseUrl,
        organization: config.organization,
        defaultModel: config.embeddingModel,
        timeout: config.timeout,
        maxRetries: config.maxRetries,
      });
      this._embeddingProviders.set(ptype, createEmbeddingServiceProvider(embeddingProvider));
    } else if (ptype === ProviderType.ANTHROPIC) {
      const llmProvider = new AnthropicLLMProvider({
        apiKey: config.apiKey,
        baseUrl: config.baseUrl,
        modelMapping: config.modelMapping,
        timeout: config.timeout,
        maxRetries: config.maxRetries,
      });
      this._llmProviders.set(ptype, createLLMServiceProvider(llmProvider));
    } else if (ptype === ProviderType.LOCAL) {
      const backend = config.baseUrl === 'fastembed' || config.baseUrl === 'sentence-transformers'
        ? config.baseUrl
        : 'fastembed';
      const embeddingProvider = new LocalEmbeddingProvider({
        modelName: config.embeddingModel,
        backend,
      });
      this._embeddingProviders.set(ptype, createEmbeddingServiceProvider(embeddingProvider));
    }
  }

  /**
   * Shutdown all services
   */
  shutdown(): void {
    if (this._cache) {
      this._cache.clear().catch(() => {});
      this._cache.close();
    }

    this._llmProviders.clear();
    this._embeddingProviders.clear();
    this._llmService = null;
    this._embeddingService = null;
    this._initialized = false;
    ServiceManager._instance = null;
  }

  /**
   * Get LLM service
   */
  get llm(): LLMServiceImpl {
    if (!this._llmService) {
      throw new Error('ServiceManager not initialized. Call initialize() first.');
    }
    return this._llmService;
  }

  /**
   * Get Embedding service
   */
  get embedding(): EmbeddingServiceImpl {
    if (!this._embeddingService) {
      throw new Error('ServiceManager not initialized. Call initialize() first.');
    }
    return this._embeddingService;
  }

  /**
   * Check all provider health statuses
   */
  async checkHealth(): Promise<Record<string, boolean>> {
    const checks: Record<string, boolean> = {};

    for (const [ptype, provider] of this._llmProviders) {
      checks[`llm_${ptype}`] = await provider.healthCheck();
    }

    for (const [ptype, provider] of this._embeddingProviders) {
      checks[`embedding_${ptype}`] = await provider.healthCheck();
    }

    this._healthStatus = checks;
    this._lastHealthCheckMs = Date.now();
    return checks;
  }

  /**
   * Check if any service is healthy
   */
  isHealthy(): boolean {
    return Object.values(this._healthStatus).some(Boolean);
  }

  /**
   * Get health status
   */
  getHealthStatus(): Record<string, boolean> {
    return { ...this._healthStatus };
  }

  /**
   * 请求失败时触发按需健康检查（带最小间隔防抖）
   * 外部调用方在 catch 块中调用此方法即可
   */
  triggerOnDemandHealthCheck(): void {
    const now = Date.now();
    if (now - this._lastHealthCheckMs < this._minHealthCheckIntervalMs) return;
    this._lastHealthCheckMs = now;
    this.checkHealth().catch(() => {
      // 按需健康检查失败仅记录日志，不中断流程
    });
  }

  /**
   * Get cache statistics
   */
  async getCacheStats(): Promise<Record<string, unknown> | null> {
    if (this._cache) {
      return this._cache.stats();
    }
    return null;
  }
}
